### ISSUE TITLE
[BUG] Unable to enable passthrough

### ISSUE THREAD

**Issue by abonillabeeche** [harvester/harvester#2973]
**Describe the bug**
I am unable to set the desired PCI Devices as allocatable

**To Reproduce**
1. Select PCI/nvidia device, including other devices on the same iommu group.
2. Enable passthrough
3. Device appears as enabled but the logs show otherwise, `Kernel driver in use: vfio-pci` won't be listed under the specific device either.

**Expected behavior**
`Kernel driver in use: vfio-pci` visible under the device and Allocatable Device under the host
**Support bundle**

[supportbundle-enablepassthrough-p5820.zip](https://github.com/harvester/harvester/files/9812251/supportbundle-enablepassthrough-p5820.zip)

**Environment**
 - Harvester ISO version:  1.1.0-rc3
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Dell Precision 5820

**Additional context**
```time="2022-10-18T15:34:28Z" level=info msg="Reconciling PCI Devices list"
time="2022-10-18T15:34:28Z" level=info msg="Attempting to disable passthrough for p5820-000065000"
time="2022-10-18T15:34:28Z" level=info msg="Attempting to disable passthrough for p5820-000065001"
time="2022-10-18T15:34:29Z" level=info msg="Attempting to disable passthrough for p5820-000065001"
time="2022-10-18T15:34:29Z" level=error msg="Error updating status for p5820-000065001: Operation cannot be fulfilled on pcideviceclaims.devices.harvesterhci.io \"p5820-000065001\": StorageError: invalid object, Code: 4, Key: /registry/devices.harvesterhci.io/pcideviceclaims/p5820-000065001, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: c2bb33ed-a876-4d0a-8d5b-c70e9ba207e8, UID in object meta: "
time="2022-10-18T15:34:29Z" level=error msg="error syncing 'p5820-000065001': handler PCIDeviceClaimOnRemove: Operation cannot be fulfilled on pcideviceclaims.devices.harvesterhci.io \"p5820-000065001\": StorageError: invalid object, Code: 4, Key: /registry/devices.harvesterhci.io/pcideviceclaims/p5820-000065001, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: c2bb33ed-a876-4d0a-8d5b-c70e9ba207e8, UID in object meta: , requeuing"
WARNING: error parsing the pci address "10000:00:02.0"
WARNING: failed to get device information for PCI address 10000:00:02.0
WARNING: error parsing the pci address "10000:00:03.0"
WARNING: failed to get device information for PCI address 10000:00:03.0



time="2022-10-18T15:34:48Z" level=info msg="Attempting to enable passthrough for p5820-000065000"
time="2022-10-18T15:34:48Z" level=info msg="Attempting to enable passthrough for p5820-000065001"
time="2022-10-18T15:34:48Z" level=info msg="Binding device p5820-000065000 [10de 1cb1] to vfio-pci"
time="2022-10-18T15:34:48Z" level=info msg="Adding p5820-000065000 to KubeVirt list of permitted devices"
time="2022-10-18T15:34:49Z" level=info msg="Binding device p5820-000065001 [10de 0fb9] to vfio-pci"
time="2022-10-18T15:34:49Z" level=info msg="Adding p5820-000065001 to KubeVirt list of permitted devices"
time="2022-10-18T15:34:49Z" level=info msg="Attempting to enable passthrough for p5820-000065000"
time="2022-10-18T15:34:49Z" level=info msg="Attempting to enable passthrough for p5820-000065001"
time="2022-10-18T15:34:50Z" level=info msg="Binding device p5820-000065000 [10de 1cb1] to vfio-pci"
time="2022-10-18T15:34:50Z" level=info msg="Adding p5820-000065000 to KubeVirt list of permitted devices"
time="2022-10-18T15:34:50Z" level=error msg="Device at address 0000:65:00.1 is not bound to driver snd_hda_intel"
time="2022-10-18T15:34:50Z" level=info msg="Binding device p5820-000065001 [10de 0fb9] to vfio-pci"
time="2022-10-18T15:34:50Z" level=info msg="Adding p5820-000065001 to KubeVirt list of permitted devices"
time="2022-10-18T15:34:50Z" level=error msg="Error updating status for p5820-000065000: Operation cannot be fulfilled on pcideviceclaims.devices.harvesterhci.io \"p5820-000065000\": the object has been modified; please apply your changes to the latest version and try again"
time="2022-10-18T15:34:51Z" level=error msg="Error updating status for p5820-000065001: Operation cannot be fulfilled on pcideviceclaims.devices.harvesterhci.io \"p5820-000065001\": the object has been modified; please apply your changes to the latest version and try again"

^C
p5820:~ # lspci | grep -i nvid
0000:65:00.0 VGA compatible controller: NVIDIA Corporation GP107GL [Quadro P1000] (rev a1)
0000:65:00.1 Audio device: NVIDIA Corporation GP107GL High Definition Audio Controller (rev a1)
p5820:~ # lspci -nn -s 0000:65:00.0 -v
0000:65:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP107GL [Quadro P1000] [10de:1cb1] (rev a1) (prog-if 00 [VGA controller])
	Subsystem: Dell Device [1028:11bc]
	Flags: bus master, fast devsel, latency 0, IRQ 11, NUMA node 0
	Memory at d7000000 (32-bit, non-prefetchable) [size=16M]
	Memory at c0000000 (64-bit, prefetchable) [size=256M]
	Memory at d0000000 (64-bit, prefetchable) [size=32M]
	I/O ports at b000 [size=128]
	Expansion ROM at 000c0000 [disabled] [size=128K]
	Capabilities: [60] Power Management version 3
	Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [78] Express Legacy Endpoint, MSI 00
	Capabilities: [100] Virtual Channel
	Capabilities: [250] Latency Tolerance Reporting
	Capabilities: [128] Power Budgeting <?>
	Capabilities: [420] Advanced Error Reporting
	Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
	Capabilities: [900] #19```

**Comment by tlehman** [comment]
That log entry for 65:00.0:

```
time="2022-10-18T15:34:50Z" level=info msg="Binding device p5820-000065000 [10de 1cb1] to vfio-pci"
```

doesn't show any errors, but from `dmesg` on the p5820, 

```
[ 6708.659672] vfio-pci: probe of 0000:65:00.0 failed with error -22
```

Which @abonillabeeche found might be related to [this](http://inbox.dpdk.org/ci/8c696743-5640-9ad9-aba4-916ca8858f64@intel.com/t/), this might be because the `iommu=pt` parameter was not set on boot.

**Comment by futuretea** [comment]
@abonillabeeche Did you meet this issue again in v1.1.0?

**Comment by tlehman** [comment]
If it only says `iommu` then it means the sed expression in `update_grub_settings` wasn't called.

It's just called unconditionally [here](https://github.com/tlehman/harvester-installer/blob/master/package/harvester-os/files/usr/sbin/harv-install#L403), there might have been an error earlier in the harv-install script that prevented `update_grub_settings` from being called.

To avoid this, I can just include all the params, even ones the CPU doesn't support, since they would no-op

**Comment by abonillabeeche** [comment]
Can the fix for this issue get a review?

**Comment by tlehman** [comment]
the impact of this is that no PCI devices can be enabled for passthrough, because the `intel_iommu=on` or `intel_amd=on` was never added

**Comment by tlehman** [comment]
this only happens if there is an error in [harv-install](https://github.com/harvester/harvester-installer/blob/ea4483407a56f95945f22071846cc16f4f3c45fa/package/harvester-os/files/usr/sbin/harv-install) before line [396](https://github.com/harvester/harvester-installer/blob/ea4483407a56f95945f22071846cc16f4f3c45fa/package/harvester-os/files/usr/sbin/harv-install#L396)

**Comment by tlehman** [comment]
I checked @abonillabeeche's two machines where this problem happened and the /oem/install/console.log files exist and don't indicate the kind of error I thought might explain this.

This means that at least for these two cases of this bug, there isn't some upstream problem in `harv-install` that prevented the `update_grub_settings` from running, since we have both logs.

**Comment by ibrokethecloud** [comment]
based on support bundle and feedback on issue, the installation is using v1.1.0-rc3.

The new kernel arguments introduced by the [PR](https://github.com/harvester/harvester-installer/pull/380) would need to be in a custom build or master once this is merged.

The current kernel boot arguments are in support bundle are:
```
Oct 18 13:43:04 localhost kernel: Command line: BOOT_IMAGE=(loop0)/boot/vmlinuz console=tty1 root=LABEL=COS_ACTIVE cos-img/filename=/cOS/active.img panic=0 net.ifnames=1 rd.cos.oemtimeout=120 rd.cos.oemlabel=COS_OEM audit=1 audit_backlog_limit=8192 iommu
```

Which would match our current v1.1.x releases.

**Comment by guangbochen** [comment]
The root cause is identified as the [harv-install](https://github.com/harvester/harvester-installer/pull/383/files#diff-6a84a5d8c064b0be0d59d2d26e27e2549fcf932c02014eb3f954bec7e780d7f7) will not be executed during the upgrade, also the sed will not apply since the `iommu` does not exist in v1.0.3 ..

**Comment by tlehman** [comment]
So the solution to this issue will be to make the upgrade path set those parameters correctly

**Comment by tlehman** [comment]
https://github.com/harvester/harvester-installer/pull/383 <-- merged

**Comment by TachunLin** [comment]
### Result (kernel parameter check)

Check the kernel parameters intel_iommu=on amd_iommu=on iommu=pt **are presented** in 

- [x] New install v1.1.1 master Harvester 
- [x] Upgrade from v1.0.3 to v1.1.1 master 
- [x] Upgrade from v1.1.0 to v1.1.1 master

1. On a new master Harvester cluster, check the kernel parameters
  ```
  harvester-node-0:~ # cat /proc/cmdline
  BOOT_IMAGE=(loop0)/boot/vmlinuz console=ttyS0 console=tty1 root=LABEL=COS_ACTIVE cos- 
  img/filename=/cOS/active.img panic=0 net.ifnames=1 rd.cos.oemtimeout=120 rd.cos.oemlabel=COS_OEM audit=1 
  audit_backlog_limit=8192 intel_iommu=on amd_iommu=on iommu=pt
  ``` 
2. Cluster upgraded from v1.0.3 to v1.1.1 master, check the kernel parameters
  ```
  harvester-node-0:~ # cat /proc/cmdline
  BOOT_IMAGE=(loop0)/boot/vmlinuz console=tty1 root=LABEL=COS_ACTIVE cos-img/filename=/cOS/active.img panic=0 
  net.ifnames=1 rd.cos.oemtimeout=120 rd.cos.oemlabel=COS_OEM audit=1 audit_backlog_limit=8192 intel_iommu=on 
  amd_iommu=on iommu=pt
  ```

3. Cluster upgrade from v1.1.0 to v1.1.1 master, check the kernel parameters
  ```
  harv110:~ # cat /proc/cmdline
  BOOT_IMAGE=(loop0)/boot/vmlinuz console=tty1 root=LABEL=COS_ACTIVE cos-img/filename=/cOS/active.img panic=0 
  net.ifnames=1 rd.cos.oemtimeout=120 rd.cos.oemlabel=COS_OEM audit=1 audit_backlog_limit=8192 intel_iommu=on 
  amd_iommu=on iommu=pt
  ```

### Test Information
* Test Environment: 4 nodes Harvester on kvm machines
* Harvester version: `master-b0d883ce-head`

### Verify Steps
1. Create a v1.1.1 master Harvester 
2. ssh to management node 
3. Run `cat /proc/cmdline`
4. Create a v1.0.3 Harvester 
5. Upgrade to v1.1.1 master 
6. Check step 3 
7. Create a v1.1.0 Harvester
8. Upgrade to v1.1.1 master 
9. Check step 3

**Comment by TachunLin** [comment]
### Result (Enable PCI passthrough on Nvidia device) 

On a new installation Harvester cluster with v1.1.1 master release.

* We can enable the Nvidia Qaudro graphic controller 
![image](https://user-images.githubusercontent.com/29251855/203736744-36d71795-c176-4abc-ba95-e9cc2892c195.png)

* Can **display** `Kernel driver in use: vfio-pci visible` under the harvester cluster 
```
node1:~ # lspci | grep -i nvid
17:00.0 VGA compatible controller: NVIDIA Corporation GP107GL [Quadro P620] (rev a1)
17:00.1 Audio device: NVIDIA Corporation GP107GL High Definition Audio Controller (rev a1)
65:00.0 VGA compatible controller: NVIDIA Corporation GP107GL [Quadro P620] (rev a1)
65:00.1 Audio device: NVIDIA Corporation GP107GL High Definition Audio Controller (rev a1)

node1:~ # lspci -nn -s 0000:17:00.0 -v
17:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP107GL [Quadro P620] [10de:1cb6] (rev a1) (prog-if 00 [VGA controller])
	Subsystem: Dell Device [1028:1264]
	Flags: fast devsel, IRQ 11, NUMA node 0
	Memory at b4000000 (32-bit, non-prefetchable) [disabled] [size=16M]
	Memory at 380060000000 (64-bit, prefetchable) [disabled] [size=256M]
	Memory at 380070000000 (64-bit, prefetchable) [disabled] [size=32M]
	I/O ports at 7000 [disabled] [size=128]
	Expansion ROM at b5000000 [disabled] [size=512K]
	Capabilities: [60] Power Management version 3
	Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [78] Express Legacy Endpoint, MSI 00
	Capabilities: [100] Virtual Channel
	Capabilities: [250] Latency Tolerance Reporting
	Capabilities: [128] Power Budgeting <?>
	Capabilities: [420] Advanced Error Reporting
	Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
	Capabilities: [900] #19
	Kernel driver in use: vfio-pci
```

### Test Information
* Test Environment: 3 nodes harvester on Dell Precise 5820 machine
* PCI devices: 6 Nvidia Qaudro P620 graphic cards ( each node have two)
* Harvester version: master-b0d883ce-head (11/22)

### Verify Steps
1. Provision a 3 nodes Harvester cluster 
2. Enable the pcidevices-controller addon
3. Go to PCI Devices page 
4. search nvidia device
5. Enable all `NVIDIA Corporation GP107GL [Quadro P620]` 
![image](https://user-images.githubusercontent.com/29251855/203774082-187b7969-fa98-4b43-b960-a6e1f6b77703.png)
6. ssh to management node 
7. Query nvidia device `lspci | grep -i nvid`
8. Check the graphic card details `lspci -nn -s 0000:17:00.0 -v`

**Comment by TachunLin** [comment]
Verified **fixed** on `master-b0d883ce-head ` (11/22) of the following check point. Close this issue.

- [x] Kernel parameter check 
- [x] Enable PCI passthrough on Nvidia device
